### ISSUE TITLE
question about transfer learning

### ISSUE THREAD

**Issue by SoodabehGhaffari** [chemprop/chemprop#899]
Hello,
I have a pretrained model using chemprop on a large dataset and I would like to use transfer learning to build a new model on a small dataset. To this end, I would like to keep the directed-message passing neural network frozen and the learned parameters of the feed forward network of the pre-trained model to initialize the parameters of the new model. This method is use in the paper below.
My question is that I need to use the two commands at the same time? Am I right?
--checkpoint_dir
--checkpoint_frzn

My another question is how I can use the ensemble of pretrained models in the transfer learning with the method that I describe above? I know the command (--checkpoint_dir) can take the ensemble of models, however, the command (--checkpoint_frzn) only take one checkpoint. 

Thank you very much for the help in advance
Best Regards
Soodabeh

**Comment by kevingreenman** [comment]
Hi @SoodabehGhaffari, I think your first question was answered in one of your previous issues, #756.

For your second question, as the README says, "At present, only one checkpoint can be provided for the --checkpoint_frzn and those parameters will be used for any number of submodels if --ensemble_size is specified." So you would need to fine tune each pretrained model separately if you want to take that approach.

**Comment by SoodabehGhaffari** [comment]
Hi,
Thanks for the prompt reply.
Let me clarify my first question:
In my previous issue #756, I wanted to keep the D-MPNN frozen and tune the
FFN. In that case, I only need to use --checkpoint_frzn. This is the answer
that I got for my previous question. However, my new question is I would
like to keep the D-MPNN frozen and initialize the FFN from the pretrained
model, so my guess is that I need to use both commands --checkpoint_dir and
--checkpoint_frzn. Is my guess right?

Thank you for the answer to my second question too.

Best Regards
Soo

**Comment by KnathanM** [comment]
>However, my new question is I would like to keep the D-MPNN frozen and initialize the FFN from the pretrained model

I believe that --checkpoint_frzn initializes both the MPNN and FFN from the pretrained model and then only freezes the MPNN (if --frzn_ffn_layers is 0). So you should be okay to only use --checkpoint_frzn. If you find this isn't the behavior of chemprop v1, please let us know.

**Comment by SoodabehGhaffari** [comment]
Thank you very much for the help!
I have one more question:
I know that the model architecture of the new model should match the old
model in any layers that are being frozen. I was wondering if we can tune
some hyperparameters that are not related to the architecture including
batch size, dropout, warmup_epochs and learning rates to improve the new
model's performance in transfer learning.

Thanks again for your help
Best Regards
Soodabeh

**Comment by SoodabehGhaffari** [comment]
I came up with another question:
How can I double check that the  Directed Message Passing Neural Network
layers are frozen and the FFN are initialized from the pretrained model and
are tuned in the new model? Is there any straightforward way that I can
check the above from the output?

Thanks again for your help
Best,
Soo

**Comment by kevingreenman** [comment]
> Thank you very much for the help! I have one more question: I know that the model architecture of the new model should match the old model in any layers that are being frozen. I was wondering if we can tune some hyperparameters that are not related to the architecture including batch size, dropout, warmup_epochs and learning rates to improve the new model's performance in transfer learning.

Yes, this could be worth trying.

> I came up with another question: How can I double check that the Directed Message Passing Neural Network layers are frozen and the FFN are initialized from the pretrained model and are tuned in the new model? Is there any straightforward way that I can check the above from the output? 

Look in the log file for "Number of unfrozen parameters" and "Total number of parameters". See [here](https://github.com/chemprop/chemprop/blob/64335d57cacd9529a08fcb2f9c671bff87067af0/chemprop/train/run_training.py#L263-L274) and [here](https://github.com/chemprop/chemprop/blob/64335d57cacd9529a08fcb2f9c671bff87067af0/chemprop/utils.py#L195) for where this is handled in the code.

**Comment by SoodabehGhaffari** [comment]
Thank you!
